olp-cpp-sdk-dataservice-read/src/repositories/CatalogRepository.cpp

```diff
@@ -35,7 +35,64 @@
 #include "olp/dataservice/read/VersionsRequest.h"
 
 namespace {
+constexpr auto kDefaultStartVersion = 0;
 constexpr auto kLogTag = "CatalogRepository";
+
+// The function retrieves the maximum version among cached/online/user-provided
+// versions.
+// Update the cache, if the retrieved version is greater than the cached one.
+olp::dataservice::read::CatalogVersionResponse RetrieveLatestVersion(
+    const olp::client::HRN& catalog,
+    const olp::dataservice::read::CatalogVersionRequest& request,
+    olp::dataservice::read::repository::CatalogCacheRepository& repository,
+    olp::dataservice::read::CatalogVersionResponse&& version_response) {
+  auto cached_version = repository.GetVersion();
+  auto fetch_option = request.GetFetchOption();
+
+  // Using `GetStartVersion` to set up a new latest version for CacheOnly
+  // requests in case of absence of the previous latest version or it's less
+  // than the new user set version.
+  if (fetch_option == olp::dataservice::read::FetchOptions::CacheOnly) {
+    const auto start_version = request.GetStartVersion();
+    if (start_version >= kDefaultStartVersion &&
+        (!cached_version || start_version > cached_version->GetVersion())) {
+      olp::dataservice::read::model::VersionResponse new_response;
+      new_response.SetVersion(start_version);
+      version_response = std::move(new_response);
+    }
+  }
+
+  if (version_response) {
+    const auto new_version = version_response.GetResult().GetVersion();
+
+    // Write or update a version in the cache, updating happens only when the
+    // new version is greater than cached.
+    if (!cached_version || cached_version->GetVersion() < new_version) {
+      repository.PutVersion(version_response.GetResult());
+      if (fetch_option == olp::dataservice::read::FetchOptions::CacheOnly) {
+        OLP_SDK_LOG_DEBUG_F(
+            kLogTag, "Latest user set version, hrn='%s', version=%" PRId64,
+            catalog.ToCatalogHRNString().c_str(), new_version);
+      } else {
+        OLP_SDK_LOG_DEBUG_F(kLogTag,
+                            "Latest online version, hrn='%s', version=%" PRId64,
+                            catalog.ToCatalogHRNString().c_str(), new_version);
+      }
+    }
+
+    return std::move(version_response);
```

Comment:
pessimizing-move
More info: https://stackoverflow.com/questions/62061433/how-to-avoid-the-pessimizing-move-warning-of-nrvo
```suggestion
    return version_response;
```